README.md
<!-- Logo -->
<h1 align="center">
  <img src="https://github.com/willtheorangeguy/Running-Calculator/blob/main/docs/images/logo.png" alt="Running Calculator">
  <br>
  Running Calculator
  <br>
</h1>

<!-- Copy -->
<h4 align="center">A running distance and speed command line interface! Measures in metric and imperial units, as well as marathons and half-marathons.</h4>

<!-- Badges -->
<div align="center">
  <!-- Stability -->
  <img alt="PyPI Build State" src="https://github.com/willtheorangeguy/Running-Calculator/actions/workflows/push-to-pypi.yml/badge.svg">
  <!-- Stability -->
  <img alt="Docker State" src="https://github.com/willtheorangeguy/Running-Calculator/actions/workflows/docker-build.yml/badge.svg">
  <!-- Stability -->
  <img alt="PyTest State" src="https://github.com/willtheorangeguy/Running-Calculator/actions/workflows/pytest.yml/badge.svg">
  <!-- Stability -->
  <img alt="Pylint State" src="https://github.com/willtheorangeguy/Running-Calculator/actions/workflows/pylint.yml/badge.svg">
  <!-- CodeQL -->
  <img alt="CodeQL State" src="https://github.com/willtheorangeguy/Running-Calculator/actions/workflows/codeql.yml/badge.svg">
  <!-- Gitleaks -->
  <img alt="Gitleaks State" src="https://github.com/willtheorangeguy/Running-Calculator/actions/workflows/gitleaks.yml/badge.svg">
  <!-- Version -->
  <img alt="GitHub Version" src="https://img.shields.io/github/v/release/willtheorangeguy/Running-Calculator">
  <!-- Issues -->
  <img alt="GitHub Issues" src="https://img.shields.io/github/issues/willtheorangeguy/Running-Calculator">
  <!-- Pull Requests -->
  <img alt="GitHub Pull Requests" src="https://img.shields.io/github/issues-pr/willtheorangeguy/Running-Calculator">
  <!-- Discord -->
  <img alt="Discord Server ID" src="https://img.shields.io/discord/950267832283848764">
  <!-- Downloads -->
  <img alt="Downloads" src="https://img.shields.io/github/downloads/willtheorangeguy/Running-Calculator/total">
  <!-- Language Count -->
  <img alt="GitHub Languages" src="https://img.shields.io/github/languages/count/willtheorangeguy/Running-Calculator">
</div>

<!-- Navigation -->
<p align="center">
  <a href="#key-features">Key Features</a> •
  <a href="#download">Download</a> •
  <a href="#how-to-use">How To Use</a> •
  <a href="#support">Support</a> •
  <a href="#contributing">Contributing</a> •
  <a href="#changelog">Changelog</a> •
  <a href="#credits">Credits & Contributors</a>
</p>

<!-- Screenshot(s) -->
![screenshot](https://github.com/willtheorangeguy/Running-Calculator/blob/main/docs/images/welcome.png)

## Key Features

* Easily input distance of a run.
* Input distance can be in marathons, half-marathons, miles, feet, kilometers and meters.
* Easily input run time.
* Displays speeds in meters per second and kilometers per hour.
* Cross platform.

## Download

You can **[download](https://github.com/willtheorangeguy/Running-Calculator/releases/latest) the source code** to run the script from the command line on Windows, macOS and Linux. **This will require [Python](https://www.python.org/downloads/).**

You can **[download](https://github.com/willtheorangeguy/Running-Calculator/releases/latest) the latest executable version** of Running Calculator for Windows. **This does not require Python.**

## How To Use

To run the application, you can use [Git and the Python Interpreter](https://github.com/willtheorangeguy/Running-Calculator/main/README.md#git), which allows you to clone and run the application, [`pip`](https://github.com/willtheorangeguy/Running-Calculator/main/README.md#pip) to create a command line application, or [Docker](https://github.com/willtheorangeguy/Running-Calculator/main/README.md#docker) to create a container of the application.

### Git

To clone and run this application, you'll need [Git](https://git-scm.com/downloads) and [Python](https://www.python.org/downloads/) installed on your computer. If you would rather not use Git, you can just download the script from GitHub above. From your command line:

```bash
# Clone this repository
$ git clone https://github.com/willtheorangeguy/Running-Calculator

# Go into the repository
$ cd Running-Calculator

# Run the CLI
$ python main.py
```

### `pip`

You can install the program from the [Python Package Index](https://pypi.org/project/Running-Calculator/) through `pip`.

```bash
# Install via pip
$ pip install running-calculator

# Run the CLI
$ running-calculator
```

### Docker

You can pull the [Docker](https://www.docker.com/) image from GitHub Packages. From your command line:

```bash
# Pull image
$ docker pull ghcr.io/willtheorangeguy/running-calculator:main

# Run container
$ docker run -i -t ghcr.io/willtheorangeguy/running-calculator:main python main.py
```

## Support

**The following units are available (can be generated by using the `help` command):**

```text
Marathons        42195m
Half Marathons   21907.5m
Miles            1069.3m
Feet             0.38m
Kilometers       1000m
Meters           1m
license          View the license file. 
exit             Exit the program.
quit             Exit the program.
```

More documentation is available in the **[Documentation](https://github.com/willtheorangeguy/Running-Calculator/tree/main/docs)** and on the **[Wiki](https://github.com/willtheorangeguy/Running-Calculator/wiki)**. If more support is required, please open a **[GitHub Discussion](https://github.com/willtheorangeguy/Running-Calculator/discussions/new)** or join our **[Discord](https://discord.gg/2eBHwHaPWp)**.

## Contributing

Please contribute using [GitHub Flow](https://guides.github.com/introduction/flow). Create a branch, add commits, and [open a pull request](https://github.com/willtheorangeguy/Running-Calculator/compare).

Please read [`CONTRIBUTING`](CONTRIBUTING.md) for details on our [`CODE OF CONDUCT`](CODE_OF_CONDUCT.md), and the process for submitting pull requests to us.

## Changelog

See the [`CHANGELOG`](CHANGELOG.md) file for details.

## Credits

This software uses the following open source packages, projects, services or websites:

<!-- Credits Table -->
<table>
  <tr>
    <th align="center"><img src="https://applets.imgix.net/https%3A%2F%2Fassets.ifttt.com%2Fimages%2Fchannels%2F2107379463%2Ficons%2Fmonochrome_large.png?w=240&h=240&s=8a19bbc158996d098e2fb18310ba7f33" width="150" height="150" alt="GitHub"/></th>
    <th align="center"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c3/Python-logo-notext.svg/182px-Python-logo-notext.svg.png" width="150" height="150" alt="PSF"/></th>
    <th align="center"><img src="https://pyinstaller.readthedocs.io/en/v4.2/_static/pyinstaller-draft1a.ico" width="150" height="150" alt="PyInstaller"/></th>
    <th align="center"><img src="https://pbs.twimg.com/profile_images/912151274551885824/sjzD5vK9_400x400.jpg" width="150" height="150" alt="Carbon"/></th>
  </tr>
  <tr>
    <td align="center">GitHub</td>
    <td align="center">Python Software Foundation</td>
    <td align="center">PyInstaller</td>
    <td align="center">Carbon</td>
  </tr>
  <tr>
    <td align="center"><a href="https://github.com/">Web</a> - <a href="https://github.com/pricing">Plans</a></td>
    <td align="center"><a href="https://www.python.org/">Web</a> - <a href="https://psfmember.org/civicrm/contribute/transact?reset=1&id=2">Donate</a></td>
    <td align="center"><a href="https://pyinstaller.readthedocs.io/en/stable/">Web</a> - <a href="https://www.pyinstaller.org/funding.html#funding-by-individuals">Donate</a></td>
    <td align="center"><a href="https://carbon.now.sh/">Web</a></td>
  </tr>
</table>

## Contributors

* [@willtheorangeguy](https://github.com/willtheorangeguy) - Sponsor on [PayPal](https://paypal.me/wvdg44?country.x=CA&locale.x=en_US)

## You may also like...

* [PyWorkout](https://github.com/willtheorangeguy/PyWorkout) - A minimal CLI to keep you inspired during your workout!
* [PyAvatar](https://github.com/willtheorangeguy/PyAvatar) - Easily display all of your creative avatars to keep them consistent across websites.
* [Python Logo Widgets](https://github.com/willtheorangeguy/Python-Logo-Widgets) - Python Powered Logo widgets that can be added to any GUI project.

## License

This project is licensed under the [GNU General Public License](https://www.gnu.org/licenses/gpl-3.0.en.html) - see the [`LICENSE`](LICENSE.md) file for details. See the [Privacy Policy](https://github.com/willtheorangeguy/Running-Calculator/blob/main/docs/legal/PRIVACY.md) and [Terms and Conditions](https://github.com/willtheorangeguy/Running-Calculator/blob/main/docs/legal/TERMS.md) for legal information.


CHANGELOG.md
# Changelog

## [v1.7.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.7.0)

### Added

- PyTest unit tests.

### Changed

- New functions for calculation and display of run distance.

## [v1.6.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.6.0)

### Changed

- Fix the Python package file.

## [v1.5.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.5.0)

### Added

- Files for a Python package.

## [v1.4.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.4.0)

### Added

- Docker container.

## [v1.3.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.3.0)

### Added

- The `license` command is now available in the program.

## [v1.2.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.2.0)

### Added

- License text to the top of each code file.

## [v1.1.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.1.0)

### Added

- `help` command option.

### Changed

- Constant variables to be named with CAPITAL letters.
- Made the code more readable.

## [v1.0.0](https://github.com/willtheorangeguy/Running-Calculator/releases/tag/v1.0.0)

### Added

- CLI


tests/test_main.py
"""This is a test file for the Running Calculator program."""
# pylint: disable=locally-disabled, wrong-import-position, import-error

import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from main import calculate_distance, display_result


def test_calculate_distance():
    """Test if calculate_distance returns the correct distance and unit."""
    # Test when distance is greater than 1000
    distance, unit = calculate_distance(1500, "meters")
    assert distance == 1.5
    assert unit == "kilometers"

    # Test when distance is less than 1000
    distance, unit = calculate_distance(500, "meters")
    assert distance == 500
    assert unit == "meters"


def test_display_result(capsys):
    """Test if display_result prints the correct output."""
    # Test when unit is "meters"
    display_result(500, "meters", 10, 36)
    captured = capsys.readouterr()
    assert "You traveled 500 meters." in captured.out
    assert (
        "Your speed was 10 meters per second or 36 kilometers per hour." in captured.out
    )

    # Test when unit is "kilometers"
    display_result(1.5, "kilometers", 10, 36)
    captured = capsys.readouterr()
    assert "You traveled 1.5 kilometers." in captured.out
    assert (
        "Your speed was 10 meters per second or 36 kilometers per hour." in captured.out
    )


Running Calculator.py
# Running Calculator Concept
# (C) 2022 - For the Sada's

# MATH CONSTANTS
ft = 3.281
miles = 1609.344
half_marathon = 21097.5
marathon = 42195

# CONVERSION VARIABLES
unit = ""

# WELCOME
run = True
length_run = True

print(r"""
                                        ,////,
                                        /// 6|
                                        //  _|
                                       _/_,-'
                                  _.-/'/   \   ,/;,
                               ,-' /'  \_   \ / _/
                               `\ /     _/\  ` /
                                 |     /,  `\_/
                                 |     \'
                     /\_        /`      /\
                   /' /_``--.__/\  `,. /  \
                  |_/`  `-._     `\/  `\   `.
                            `-.__/'     `\   |
                                          `\  \
                                            `\ \
                                              \_\__
                                               \___)""")
print(r"""
                                        _             
                                       (_)            
                 _ __ _   _ _ __  _ __  _ _ __   __ _ 
                | '__| | | | '_ \| '_ \| | '_ \ / _` |
                | |  | |_| | | | | | | | | | | | (_| |
                |_|   \__,_|_| |_|_| |_|_|_| |_|\__, |
                                                 __/ |
                                                |___/ 
""")
print("              WELCOME TO THE RUNNING SPEED CALCULATOR \n    Please answer each question to receive your time and speed! \nNote: Currently, the program only displays results in metric values. \n")

# LENGTH ENTRY
# Type
while run == True:
    while length_run == True:
        length_type = str(input("What is the name of unit you will be inputting? "))
        if length_type.lower() == "marathons":
            unit="marathon"
            print("You are now entering in marathons! \n")
            length_run = False
        elif length_type.lower() == "half marathons":
            unit="halfmarathon"
            print("You are now entering in half-marathons! \n")
            length_run = False
        elif length_type.lower() == "miles":
            unit="miles"
            print("You are now entering in miles! \n")
            length_run = False
        elif length_type.lower() == "feet":
            unit="feet"
            print("You are now entering in feet! \n")
            length_run = False
        elif length_type.lower() == "kilometers":
            unit="kilometers"
            print("You are now entering in kilometers! \n")
            length_run = False
        elif length_type.lower() == "meters":
            unit="meters"
            print("You are now entering in meters! \n")
            length_run = False
        elif length_type.lower() == "quit":
            length_run = False
            run = False 
            exit()
        else:
            length_run = True
            print("Please try again. Options are: \n - Marathons \n - Half Marathons \n - Miles \n - Feet  \n - Kilometers \n - Meters \n - Quit (to exit) \n")

    # Value - needs to be in meters
    distance = 0
    length = float(input("How far did you run? "))
    hours = float(input("How many hours did it take you? (if < 1, enter 0) "))
    mins = float(input("How many minutes did it take you? (if < 1, enter 0) "))
    secs = float(input("How many seconds did it take you? (if < 1, enter 0) "))

    # DISTANCE CALCULATION
    if unit == "marathon":
        distance = length*marathon
        unit == "meters"
    elif unit == "halfmarathon":
        distance = length*half_marathon
        unit == "meters"
    elif unit == "miles":
        distance = length*miles
        unit == "meters"
    elif unit == "feet":
        distance = length/ft
        unit == "meters"
    elif unit == "kilometers":
        distance = length*1000
        unit == "meters"
    elif unit == "meters":
        distance = length
        unit == "meters"
    else:
        pass

    # TIME CALCULATION
    time = 0
    time = (hours*3600) + (mins*60) + secs
    
    # SPEED CALCULATION
    metersps = distance/time
    kmph = metersps*3.6

    # PRINT FINAL RESPONSES
    # Change very long meters to kms
    if distance >= 1000:
        distance = distance/1000
        unit = "kilometers"
    else:
        distance = distance
        unit = "meters"

    print("")
    if unit == "meters":
        print("You travelled " + str(round(distance, 2)) + " meters.")
    else:
        print("You travelled " + str(round(distance, 2)) + " kilometers.")
    print("Your speed was " + str(round(metersps, 2)) + " meters per second or " + str(round(kmph, 2)) + " kilometers per hour. \n\n")
    length_run = True 


main.py
"""
Running Calculator Concept - for the Sada's
Copyright (C) 2022-2023 @willtheorangeguy

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 3 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
# pylint: disable=locally-disabled, invalid-name, too-many-locals, too-many-branches, too-many-statements

# Running Calculator
import sys


def calculate_distance(distance, unit):
    """Calculate the distance in meters."""
    if distance > 1000:
        distance = distance / 1000
        unit = "kilometers"
    else:
        unit = "meters"
    return distance, unit


def display_result(distance, unit, metersps, kmph):
    """Display the result."""
    print("")
    if unit == "meters":
        print("You traveled " + str(round(distance, 2)) + " meters.")
    else:
        print("You traveled " + str(round(distance, 2)) + " kilometers.")
    print(
        "Your speed was "
        + str(round(metersps, 2))
        + " meters per second or "
        + str(round(kmph, 2))
        + " kilometers per hour. \n\n"
    )


def main():
    """Main entry point of the app."""
    # MATH CONSTANTS
    FT = 3.281
    MILES = 1609.344
    HALF_MARATHON = 21097.5
    MARATHON = 42195

    # CONVERSION VARIABLES
    unit = ""

    # WELCOME
    run = True
    length_run = True
    time_run = True

    print(
        r"""
                                           ,////,
                                           /// 6|
                                           //  _|
                                          _/_,-'
                                     _.-/'/   \   ,/;,
                                  ,-' /'  \_   \ / _/
                                  `\ /     _/\  ` /
                                    |     /,  `\_/
                                    |     \'
                        /\_        /`      /\
                      /' /_``--.__/\  `,. /  \
                     |_/`  `-._     `\/  `\   `.
                               `-.__/'     `\   |
                                             `\  \
                                               `\ \
                                                 \_\__
                                                  \___)"""
    )
    print(
        r"""
                                        _
                                        (_)
                    _ __ _   _ _ __  _ __  _ _ __   __ _
                | '__| | | | '_ \| '_ \| | '_ \ / _` |
                | |  | |_| | | | | | | | | | | | (_| |
                |_|   \__,_|_| |_|_| |_|_|_| |_|\__, |
                                                 __/ |
                                                |___/
    """
    )
    print("              WELCOME TO THE RUNNING SPEED CALCULATOR")
    print("    Please answer each question to receive your time and speed!")
    print("Note: Currently, the program only displays results in metric values. \n")

    # LENGTH ENTRY
    while run:
        # Type of Measurement
        while length_run:
            length_type = str(input("What unit will you be inputting? "))
            if length_type.lower() == "marathons" or length_type.lower() == "marathon":
                unit = "marathon"
                print("You are now entering in marathons! \n")
                length_run = False
            elif (
                length_type.lower() == "half marathons"
                or length_type.lower() == "half marathon"
            ):
                unit = "halfmarathon"
                print("You are now entering in half-marathons! \n")
                length_run = False
            elif length_type.lower() == "miles" or length_type.lower() == "mile":
                unit = "miles"
                print("You are now entering in miles! \n")
                length_run = False
            elif length_type.lower() == "feet" or length_type.lower() == "foot":
                unit = "feet"
                print("You are now entering in feet! \n")
                length_run = False
            elif (
                length_type.lower() == "kilometers"
                or length_type.lower() == "kilometer"
            ):
                unit = "kilometers"
                print("You are now entering in kilometers! \n")
                length_run = False
            elif length_type.lower() == "meters" or length_type.lower() == "meter":
                unit = "meters"
                print("You are now entering in meters! \n")
                length_run = False
            elif length_type.lower() == "license":
                print("Running Calculator Copyright (C) 2022-2023 @willtheorangeguy")
                print(
                    "This program comes with ABSOLUTELY NO WARRANTY;"
                    " for details view the license."
                )
                print("This is free software, and you are welcome to redistribute it")
                print("under certain conditions; view the license for details. \n")
                length_run = True
            elif length_type.lower() == "quit" or length_type.lower() == "exit":
                length_run = False
                run = False
                sys.exit()
            else:
                print("Please try again. Options are:")
                print("Marathons\t 42195m")
                print("Half Marathons\t 21907.5m")
                print("Miles\t\t 1069.3m")
                print("Feet\t\t 0.38m")
                print("Kilometers\t 1000m")
                print("Meters\t\t 1m")
                print("license\t\t View the license file.")
                print("exit\t\t Exit the program. \n")
                print("quit\t\t Exit the program. \n")
                length_run = True

        # Value
        distance = 0
        while time_run:
            try:  # enable only numerical values
                time_run = False
                length = float(input("How far did you run? "))
                hours = int(input("How many hours did it take you? (if < 1, enter 0) "))
                mins = int(
                    input("How many minutes did it take you? (if < 1, enter 0) ")
                )
                secs = float(input("How many seconds did it take you? "))
            except ValueError:
                print(
                    "Values must be in numerical form, "
                    "and only seconds can be a decimal! \n"
                )
                time_run = True

        # DISTANCE CALCULATION - conversion to meters
        if unit == "marathon":
            distance = length * MARATHON
        elif unit == "halfmarathon":
            distance = length * HALF_MARATHON
        elif unit == "miles":
            distance = length * MILES
        elif unit == "feet":
            distance = length / FT
        elif unit == "kilometers":
            distance = length * 1000
        elif unit == "meters":
            distance = length
        else:
            pass

        # TIME CALCULATION
        time = 0
        time = (hours * 3600) + (mins * 60) + secs

        # SPEED CALCULATION
        metersps = distance / time
        kmph = metersps * 3.6

        # PRINT FINAL RESPONSES
        distance, unit = calculate_distance(distance, unit)
        display_result(distance, unit, metersps, kmph)

        length_run = True
        time_run = True


# Execute only if run as the entry point into the program
if __name__ == "__main__":
    main()
